runtime/src/main/scala/akka/grpc/internal/GrpcProtocolWeb.scala

```diff
@@ -7,33 +7,50 @@ package akka.grpc.internal
 import akka.grpc.GrpcProtocol._
 import akka.http.scaladsl.model._
 import akka.http.scaladsl.model.HttpEntity.{ Chunk, ChunkStreamPart }
-import akka.util.ByteString
+import akka.util.{ ByteString, ByteStringBuilder }
 import io.grpc.{ Status, StatusException }
 
+import scala.collection.immutable
+
 abstract class GrpcProtocolWebBase(subType: String) extends AbstractGrpcProtocol(subType) {
   protected def postEncode(frame: ByteString): ByteString
   protected def preDecode(frame: ByteString): ByteString
 
   override protected def writer(codec: Codec): GrpcProtocolWriter =
-    AbstractGrpcProtocol.writer(this, codec, frame => encodeFrame(codec, frame), encodeDataToFrameBytes(codec, _))
+    AbstractGrpcProtocol.writer(this, codec, frame => encodeFrame(codec, frame), encodeDataToResponse(codec))
 
   override protected def reader(codec: Codec): GrpcProtocolReader =
     AbstractGrpcProtocol.reader(codec, decodeFrame, preDecode)
 
-  @inline
-  private def encodeFrame(codec: Codec, frame: Frame): ChunkStreamPart = {
-    val framed = frame match {
+  private def encodeFrame(codec: Codec, frame: Frame): ChunkStreamPart =
+    Chunk(postEncode(encodeFrameToBytes(codec, frame)))
+
+  private def encodeDataToResponse(
+      codec: Codec)(data: ByteString, headers: immutable.Seq[HttpHeader], trailer: Trailer): HttpResponse =
+    HttpResponse(
+      status = StatusCodes.OK,
+      headers = headers,
+      entity = HttpEntity(contentType, encodeDataToFrameBytes(codec, data, trailer)),
+      protocol = HttpProtocols.`HTTP/1.1`)
+
+  private def encodeDataToFrameBytes(codec: Codec, data: ByteString, trailer: Trailer): ByteString = {
+    val trailerData = encodeTrailerHeaders(trailer.headers.iterator)
+    val trailerFrame =
+      AbstractGrpcProtocol.encodeFrameData(codec.compress(trailerData), codec.isCompressed, isTrailer = true)
+    postEncode(encodeFrameToBytes(codec, DataFrame(data)) ++ trailerFrame)
+  }
+
+  private def encodeFrameToBytes(codec: Codec, frame: Frame): ByteString =
+    frame match {
       case DataFrame(data) =>
         AbstractGrpcProtocol.encodeFrameData(codec.compress(data), codec.isCompressed, isTrailer = false)
       case TrailerFrame(trailer) =>
-        AbstractGrpcProtocol.encodeFrameData(encodeTrailer(trailer), codec.isCompressed, isTrailer = true)
```

Comment:
This line might be another bug I introduced in the optimization: we potentially report compression but don't actually do compression.